### PR TITLE
fmt: Update for Conan V2

### PR DESCRIPTION
**Pull request by jwillikers**
Specify library name and version:  **fmt/***

closes https://github.com/conan-io/conan-center-index/pull/12162
fixes https://github.com/conan-io/conan/issues/11870

---

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/fmt/all/conanfile.py [hunk 1]
```diff
@@ -42,13 +41,19 @@ def _has_with_os_api_option(self):
     def generate(self):
         if not self.options.header_only:
             tc = CMakeToolchain(self)
-            tc.generate()  
+            tc.cache_variables["FMT_DOC"] = False
+            tc.cache_variables["FMT_TEST"] = False
+            tc.cache_variables["FMT_INSTALL"] = True
+            tc.cache_variables["FMT_LIB_DIR"] = "lib"
+            if self._has_with_os_api_option:
+                tc.cache_variables["FMT_OS"] = bool(self.options.with_os_api)
```
Comment (SpaceIm) [inline recipes/fmt/all/conanfile.py:49]: no need for cache_variables, variables is fine, except maybe for `FMT_LIB_DIR`.

Comment (jwillikers) [reply recipes/fmt/all/conanfile.py:49]: What about the previous comment that was in the recipe?

```
# FIXME : https://github.com/conan-io/conan/issues/11476
# can be replaced by https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmaketoolchain.html#cache-variables in 1.50
```

Comment (uilianries) [reply recipes/fmt/all/conanfile.py:49]: variables is fine, you are passing defines to build fmt only. Replace cache_variable, so we can merge it soon.

Comment (SSE4) [reply recipes/fmt/all/conanfile.py:49]: I think fmt is already both v1 and v2 compatible, at least we were able to build it in both environments sucessfully

File: recipes/fmt/all/conanfile.py [hunk 2]
```diff
@@ -42,13 +41,19 @@ def _has_with_os_api_option(self):
...
+            tc.generate()
 
     def layout(self):
-        cmake_layout(self)
+        cmake_layout(self, src_folder="src")
 
     def config_options(self):
-        if self.settings.os == "Windows":
+        if self.settings.os == "Windows" or self.options.shared:
```
Comment (uilianries) [inline recipes/fmt/all/conanfile.py:56]:
```suggestion
        if self.settings.os == "Windows":
```
Config options does not parse options.

File: recipes/fmt/all/conanfile.py (hunk 1 above)
Comment (uilianries) [reply recipes/fmt/all/conanfile.py:49]: partially compatible, yes. This PR improves V2 compatibility. `self.info.header_only()` is now `self.info.clear()`. And, builddirs should be empty.

Comment (uilianries) [reply recipes/fmt/all/conanfile.py:49]: @jwillikers don't forget to replace `cache_variables` by `variables` please

Comment (jwillikers) [reply recipes/fmt/all/conanfile.py]: @uilianries Thanks for the mention, I missed the responses to this thread. I've fixed that now.

**Review by uilianries — APPROVED** [review]
LGTM

**Review by uilianries — APPROVED** [review]

**Review by jgsogo — APPROVED** [review]

**Pull request merged**